### PR TITLE
Phase 6: Packed Ensemble — M Sub-Models in One Forward Pass

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis
Researcher-agent's #1 idea: Packed Ensembles embed M sub-models inside one forward pass using grouped linear operations. Same variance reduction as deep ensembles but 25% faster training (one model trains all sub-models simultaneously). Reference: Laurent et al. arXiv:2312.13403.

Implementation: Replace each nn.Linear(in, out) with nn.Linear(in*M, out*M) using groups=M. Train once, get M predictions. Average at inference.

| GPU 0-1 | Baseline s42/s43 |
| GPU 2-3 | Packed M=2 s42/s43 |
| GPU 4-5 | Packed M=4 s42/s43 |
| GPU 6-7 | Packed M=8 s42/s43 |

`python train.py --agent thorfinn --wandb_group phase6/packed-ensemble --asinh_pressure --asinh_scale 0.75` + all baseline flags

Baseline ensemble: p_in 12.4, p_oodc 6.7, p_tan 29.4, p_re 5.8

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT thorfinn:

## Results

### Implementation

Packed Ensembles (Laurent et al., ICLR 2023): M independent sub-networks in one forward pass via grouped linear operations. Each TransolverBlock's MLP and SE layers use `PackedLinear` (block-diagonal weights, groups=M). The last block returns hidden features to M independent output heads (each sees hidden_dim/M features). Predictions are averaged across M sub-models. Attention remains shared (provides common context). Alpha=1 (parameter-matched total; each sub-model is 1/M as wide as baseline).

### Experiment Setup

All runs use baseline flags + `--asinh_pressure --asinh_scale 0.75` + all baseline flags from BASELINE.md. Two seeds per M value (42, 43). W&B group: `phase6/packed-ensemble`.

| GPU | Config | Seed | Epochs | p_in | p_oodc | p_tan | p_re | val/loss | VRAM |
|-----|--------|------|--------|------|--------|-------|------|----------|------|
| 0 | **Baseline** | 42 | 155 | **12.9** | **8.0** | **30.4** | **6.2** | 0.3862 | 38.0GB |
| 1 | **Baseline** | 43 | 156 | 13.5 | 8.1 | 30.7 | 6.6 | 0.3920 | 38.0GB |
| 2 | **Packed M=2** | 42 | 153 | **12.7** | 8.1 | **30.2** | 6.4 | 0.3897 | 37.6GB |
| 3 | **Packed M=2** | 43 | 152 | 13.4 | **8.0** | 30.4 | 6.6 | 0.3958 | 37.6GB |
| 4 | **Packed M=4** | 42 | 150 | 12.9 | 8.4 | **30.0** | 6.9 | 0.3998 | 37.6GB |
| 5 | **Packed M=4** | 43 | 151 | 13.8 | 8.1 | 30.3 | 6.7 | 0.4020 | 37.6GB |
| 6 | **Packed M=8** | 42 | 145 | 13.5 | 8.3 | **29.9** | 6.8 | 0.4016 | 37.7GB |
| 7 | **Packed M=8** | 43 | 146 | 13.4 | 8.5 | 30.4 | 6.9 | 0.4048 | 37.7GB |

### 2-Seed Averages

| Config | p_in (avg) | p_oodc (avg) | p_tan (avg) | p_re (avg) | val/loss (avg) |
|--------|-----------|-------------|------------|-----------|---------------|
| **Baseline** | **13.2** | **8.1** | **30.6** | **6.4** | **0.3891** |
| Packed M=2 | 13.1 | 8.1 | 30.3 | 6.5 | 0.3928 |
| Packed M=4 | 13.4 | 8.3 | 30.2 | 6.8 | 0.4009 |
| Packed M=8 | 13.5 | 8.4 | 30.2 | 6.9 | 0.4032 |
| **8-seed ensemble baseline** | **12.4** | **6.7** | **29.4** | **5.8** | — |

### W&B Run IDs

- Baseline s42: `2712o8vx`, s43: `bakapi0o`
- Packed M=2 s42: `6lu6vpn1`, s43: `wdzstqgs`
- Packed M=4 s42: `y3l7r1ay`, s43: `1enz82qf`
- Packed M=8 s42: `kys6ywq3`, s43: `gwkp71ku`

### Command

```bash
cd cfd_tandemfoil && python train.py --agent thorfinn \
  --wandb_name "thorfinn/<description>" --wandb_group phase6/packed-ensemble \
  --asinh_pressure --asinh_scale 0.75 \
  --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp \
  --n_layers 3 --slice_num 96 --tandem_ramp \
  --domain_layernorm --domain_velhead --ema_decay 0.999 \
  --weight_decay 5e-5 --cosine_T_max 160 --disable_pcgrad \
  --pressure_first --pressure_deep --residual_prediction \
  --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --packed_M <M> --seed <seed>
```

### What happened

**Packed ensembles with alpha=1 do not improve over baseline.** The results are neutral to slightly worse:

1. **M=2**: Nearly identical to baseline on all metrics. p_in improved slightly (13.1 vs 13.2), p_tan slightly better (30.3 vs 30.6), but p_re slightly worse (6.5 vs 6.4). Within seed-to-seed variance.

2. **M=4**: Moderately worse on p_oodc (+2.5%), p_re (+6.3%). p_tan slightly better (30.2 vs 30.6). val/loss 3.0% worse.

3. **M=8**: Worst performance. p_oodc +3.7%, p_re +7.8% worse than baseline. p_tan slightly better (30.2 vs 30.6, but within noise). val/loss 3.6% worse.

**Why**: At alpha=1, each sub-model is 1/M as wide as baseline. The ensemble averaging of M narrower sub-models does not compensate for the reduced per-model capacity. The shared attention provides common representations, but the grouped MLPs and independent output heads operate on too few features (e.g., M=8 → 24 features per sub-model with hidden_dim=192). The diversity from independent init is insufficient to overcome the capacity bottleneck.

**p_tan paradox**: All packed variants show marginally better p_tan (tandem transfer). This might be because narrower models have an implicit regularization effect that helps generalization on the tandem domain. But the improvement is small and within noise.

**Memory**: Slightly less than baseline (~37.6GB vs 38.0GB), as expected with alpha=1 reducing the MLP parameter count.

### Suggested follow-ups

1. **Packed M=2 with alpha=2**: Double the hidden dimension for packed layers so each sub-model maintains full baseline capacity. This is the true packed ensemble comparison (same per-model capacity as deep ensemble). Would increase VRAM to ~50-60GB.

2. **Packed output heads only (no packed MLP)**: Keep the full-width MLP but use M independent output heads on the shared backbone. Tests whether output-level diversity alone helps.

3. **Post-hoc ensemble of the packed M=2 runs**: Average the predictions from the M=2 s42 and s43 runs (4 effective sub-models total). This might beat the 2-model post-hoc ensemble baseline.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2082 because ADVISOR: Closing — Packed Ensemble doesn't work at our model scale.

**W&B analysis (phase6/packed-ensemble, 8 finished):**

| Config | p_in | p_oodc | p_tan | p_re |
|--------|------|--------|-------|------|
| baseline | 13.21 | 8.08 | 30.52 | 6.40 |
| M=2 | 13.06 (-1.1%) | 8.04 (-0.5%) | 30.30 (-0.7%) | 6.49 (+1.5%) |
| M=4 | 13.34 (+0.9%) | 8.26 (+2.2%) | 30.18 (-1.1%) | **6.81 (+6.4% WORSE)** |
| M=8 | 13.49 (+2.1%) | 8.37 (+3.6%) | 30.15 (-1.2%) | **6.85 (+7.1% WORSE)** |

**Root cause:** Our 1.7M-parameter model is too small for packed ensembles. M=4 splits n_hidden=192 into 4 sub-models of 48 dims each — far too narrow for the physics representation. M=8 is even worse. M=2 is marginal but doesn't outperform simply training 2 separate models.

Packed Ensembles work when the base model is large enough that each sub-model retains sufficient representational capacity. At our scale, training separate models and averaging at inference (our current approach) is strictly better.

Reassigning you to a new experiment..
